### PR TITLE
adds functionality to wait for updates being complete

### PR DESCRIPTION
**Pull request by xh3b4sd**
Towards https://github.com/giantswarm/giantswarm/issues/1918.

### CODE REVIEW

File: Gopkg.toml [hunk 1]
```diff
@@ -45,6 +45,10 @@ required = [
   branch = "master"
   name = "github.com/giantswarm/apprclient"
 
+[[override]]
+  branch = "permanent"
+  name = "github.com/giantswarm/backoff"
```
Comment (xh3b4sd) [inline Gopkg.toml]: See https://github.com/giantswarm/backoff/pull/9.

File: update/provider/aws.go [hunk 2]
```diff
@@ -6,31 +6,27 @@ import (
 	"github.com/giantswarm/e2e-harness/pkg/framework"
 	"github.com/giantswarm/microerror"
 	"github.com/giantswarm/micrologger"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/types"
 )
 
 type AWSConfig struct {
-	GuestFramework *framework.Guest
```
Comment (rossf7) [inline update/provider/aws.go]: <3

**Review by rossf7 — APPROVED** [review]
LGTM, very nice.

**Review by TheoBrigitte — APPROVED** [review]
LGTM

File: update/provider/aws.go (hunk 2 above)
Comment (indyfree) [reply update/provider/aws.go]: We didn't even use it before?

**Review by indyfree — APPROVED** [review]
Nice

File: update/provider/aws.go (hunk 2 above)
Comment (xh3b4sd) [reply update/provider/aws.go]: The test wasn't used before. Back then I just set it up based on a gut feeling to have some initial structure to iterate on. I copied some other test and didn't polish everything. Now I noticed the guest framework is not necessary for the update test.

File: update/update.go [hunk 3]
```diff
@@ -39,6 +51,33 @@ func New(config Config) (*Update, error) {
...
+	{
+		u.logger.LogCtx(ctx, "level", "debug", "message", "waiting for the guest cluster to be created")
+
+		o := func() error {
+			s, err := u.provider.CurrentStatus()
+			if err != nil {
+				return microerror.Mask(err)
+			}
+			if s.HasCreatedCondition() {
+				return backoff.Permanent(microerror.Mask(hasDesiredStatusError))
```
Comment (kopiczko) [inline update/update.go:63]: Why not just returning `nil` here?

File: update/update.go [hunk 4]
```diff
@@ -83,14 +122,57 @@ func (u *Update) Test(ctx context.Context) error {
...
-		u.logger.LogCtx(ctx, "level", "debug", "message", "updating the guest cluster with the new version bundle version")
+		u.logger.LogCtx(ctx, "level", "debug", "message", "waiting for the guest cluster to be updating")
+
+		o := func() error {
+			s, err := u.provider.CurrentStatus()
+			if err != nil {
+				return microerror.Mask(err)
+			}
+			if s.HasUpdatingCondition() {
+				return backoff.Permanent(microerror.Mask(hasDesiredStatusError))
```
Comment (kopiczko) [inline update/update.go:133]: And here.

File: update/update.go [hunk 5]
```diff
@@ -83,14 +122,57 @@ func (u *Update) Test(ctx context.Context) error {
...
+	{
+		u.logger.LogCtx(ctx, "level", "debug", "message", "waiting for the guest cluster to be updated")
+
+		o := func() error {
+			s, err := u.provider.CurrentStatus()
+			if err != nil {
+				return microerror.Mask(err)
+			}
+			if s.HasUpdatedCondition() {
+				return backoff.Permanent(microerror.Mask(hasDesiredStatusError))
```
Comment (kopiczko) [inline update/update.go:160]: And here.

File: update/update.go (hunk 3 above)
Comment (xh3b4sd) [reply update/update.go:63]: Oh wow. Captain obvious. I will provide a PR. Thanks.

Comment (xh3b4sd) [reply update/update.go:63]: See https://github.com/giantswarm/e2etests/pull/9.

**Pull request merged**